src/tmpfiles/tmpfiles.c

```diff
@@ -1547,42 +1548,71 @@ static int create_directory_or_subvolume(Item *i, const char *path) {
                          * specifically set up a btrfs subvolume for the root
                          * dir too. */
 
-                        r = -ENOTTY;
+                        subvol = false;
                 else {
-                        RUN_WITH_UMASK((~i->mode) & 0777)
-                                r = btrfs_subvol_make(i->path);
+                        RUN_WITH_UMASK((~mode) & 0777)
+                                r = btrfs_subvol_make_fd(pfd, basename(path));
                 }
         } else
                 r = 0;
 
-        if (IN_SET(i->type, CREATE_DIRECTORY, TRUNCATE_DIRECTORY) || r == -ENOTTY)
+        if (!subvol || r == -ENOTTY)
                 RUN_WITH_UMASK(0000)
-                        r = mkdir_label(i->path, i->mode);
+                        r = mkdirat_label(pfd, basename(path), mode);
 
         if (r < 0) {
                 int k;
 
                 if (!IN_SET(r, -EEXIST, -EROFS))
-                        return log_error_errno(r, "Failed to create directory or subvolume \"%s\": %m", i->path);
+                        return log_error_errno(r, "Failed to create directory or subvolume \"%s\": %m", path);
 
-                k = is_dir(i->path, false);
+                k = is_dir(path, false);
                 if (k == -ENOENT && r == -EROFS)
-                        return log_error_errno(r, "%s does not exist and cannot be created as the file system is read-only.", i->path);
+                        return log_error_errno(r, "%s does not exist and cannot be created as the file system is read-only.", path);
                 if (k < 0)
-                        return log_error_errno(k, "Failed to check if %s exists: %m", i->path);
-                if (!k) {
-                        log_warning("\"%s\" already exists and is not a directory.", i->path);
-                        return 0;
-                }
+                        return log_error_errno(k, "Failed to check if %s exists: %m", path);
+                if (!k)
+                        return log_warning_errno(EEXIST, "\"%s\" already exists and is not a directory.", path);
 
                 creation = CREATION_EXISTING;
         } else
                 creation = CREATION_NORMAL;
 
-        log_debug("%s directory \"%s\".", creation_mode_verb_to_string(creation), i->path);
+        log_debug("%s directory \"%s\".", creation_mode_verb_to_string(creation), path);
+
+        return openat(pfd, basename(path), O_NOCTTY|O_CLOEXEC|O_DIRECTORY);
```

Comment:
this isn't right. openat() returns errors in errno, but our usual coding style says errors should be return as negative values, hence needs to be changed to:

```c
r = openat(pfd, …);
if (r < 0)
        return -errno;
return 0;
```

Reply:
fixed (except that it should return the fd returned by openat() instead of 0 in case of success).